### PR TITLE
Interface robustness improvments

### PR DESCRIPTION
**Pull request by mloubout**
Very small PR that removes the assumption on the order of the dimensions in `TimeData` and `DenseData` thagt was hard-coded to `t,x,y,z`

### CODE REVIEW

**Comment by mloubout** [comment]
This fixes part of the problems of #337

File: devito/dse/inspection.py [hunk 1]
```diff
@@ -21,7 +21,10 @@ def terminals(expr, discard_indexed=False):
 
     # Use '.name' for quickly checking uniqueness
     junk = flatten([i.free_symbols for i in indexed])
-    junk = [i.name for i in junk]
+    junk = [i.name for i in junk] + \
+           [i.parent.name if
+            (i.is_Buffered if hasattr(i, "is_Buffered") else False)
```
Comment (FabioLuporini) [inline devito/dse/inspection.py]: ``... + [i.parent.name for i in junk if isinstance(i, BufferedDimension)]``

Comment (mloubout) [reply devito/dse/inspection.py]: Thanks

Just not sure if this should be here. I hacked it because it work but seems odd and out of place.

**Comment by FabioLuporini** [comment]
OK this needs to be rebased on latest master (many things can now be dropped, I think, as part of the latest bug fixes in master). Then we can start thinking about how to improve it

**Comment by mloubout** [comment]
As agreed with @FabioLuporini I will put back mandatory time as first dimension as the loop scheduling is too messy to handle with random dimensions ordering

**Comment by mlange05** [comment]
Ok, for discussion's sake, I'm not a fan of hard-coding all the "space dimensions" specifically either, but as long as we have no clear abstractions to separate Time/Space/Misc dimensions there is little we can do. This sort of separation has a long dependency chain as well though, since it directly affects how we sanity-check and default things should we ever build a generic "grid" abstraction. In fact, I think the two features are very tightly coupled, since a grid can only ever provide/check/know about "space dimensions", while things like time and misc should be optional additions for individual fields. Just my 2 cents' worth though...

File: tests/test_operator.py [hunk 2]
```diff
@@ -569,7 +569,7 @@ def test_equations_mixed_densedata_timedata(self, shape, dimensions):
         assert len(trees) == 2
         assert all(trees[0][i] is trees[1][i] for i in range(3))
 
-        # Verify both operators produce the same result
+        # Verify both operators produce the same resultsx
```
Comment (FabioLuporini) [inline tests/test_operator.py]: typo

File: devito/interfaces.py [hunk 3]
```diff
@@ -694,13 +693,24 @@ def __init__(self, *args, **kwargs):
                             'dimension (save=False). This value will be ignored!'
                             % self.name)
                 time_dim = self.time_order + 1
-                self.indices[0].modulo = time_dim
+                # Get time index
+                _t = [var for var in self.indices if var in (t, time)][0]
```
Comment (FabioLuporini) [inline devito/interfaces.py]: can we assume this is the outermost left dimension or not?

File: tests/test_operator.py (hunk 2 above)
Comment (mloubout) [reply tests/test_operator.py]: Darn

**Comment by FabioLuporini** [comment]
> which is why there is a new test.

I cannot find it

File: devito/interfaces.py (hunk 3 above)
Comment (mloubout) [reply devito/interfaces.py]: For our purpose yes, but I think it is reductive to inforce it, so I left it as a possibility to put the dimensions in any order with a warning stating you need to know what you are doing, as in some case, the order may not matter

Comment (FabioLuporini) [reply devito/interfaces.py]: One seriously big (purely technical) issue is that all those ``if t/time is not indices/dimensions then ...``  make the code a mess

**Comment by mloubout** [comment]
>This PR is touching a piece of code that was already quite awful

Let's make it clear, this piece of code is not supposed to be pretty CS, it is supposed to be the math link between Devito and whatever people want to do with it. 

> What features are you exactly trying to get

Not enforcing our specific setup to any PDE people may want to solve

>WHy do you need a time dimension different than t/time

I don't there is just no way to know wether it's `time` or `t`, we can force it to be one or the other

>this needs to be rebased on master

It;s like 5 times less commits than most of you guys PR, common

**Comment by FabioLuporini** [comment]
> it is supposed to be the math link between Devito and whatever people want to do with it

so it's even more important

> Not enforcing our specific setup to any PDE people may want to solve

Concrete examples would be useful so that I can understand what generalisation you are looking for. Is it just to allow people to use dims different than ``x/y/z`` ?

**Comment by FabioLuporini** [comment]
> It;s like 5 times less commits than most of you guys PR, common

It's not that. Being on top of master makes the ``diff`` simpler to read

File: devito/interfaces.py [hunk 4]
```diff
@@ -716,10 +726,16 @@ def _indices(cls, **kwargs):
...
-        tidx = time if save else t
-        _indices = DenseData._indices(**kwargs)
-        return tuple([tidx] + list(_indices))
+        # Get dimensions
+        dimensions = kwargs.get('dimensions', None)
+        # Add time index if not specified
+        if dimensions is None or (t not in dimensions and time not in dimensions):
+            save = kwargs.get('save', None)
+            tidx = time if save else t
+            _indices = DenseData._indices(**kwargs)
```
Comment (FabioLuporini) [inline devito/interfaces.py]: does ``super(TimeData, self)._indices(**kwargs)`` work ?

Comment (mloubout) [reply devito/interfaces.py]: This was there before, shows as diff because of the if

**Comment by mloubout** [comment]
>Is it just to allow people to use dims different than x/y/z
no just to allow any order including time

>so it's even more important
It's important to be readable for non-CS people, not fancy

**Comment by mloubout** [comment]
>It's not that. Being on top of master makes the diff simpler to read

only diff is `interfaces` and the laplacian in the test, waste of time for the exact same diff

**Comment by FabioLuporini** [comment]
> It's important to be readable for non-CS people, not fancy

Yeah, I should stop making observations only to make the code fancier :)

**Comment by FabioLuporini** [comment]
@mloubout and I discussed about this offline and we agreed this needs first a wider discussion with the rest of the development team to be done properly

**Comment by mloubout** [comment]
Needs a fresh start from recent updates, closing.

**Pull request closed without merging**